tests/test_hardware_efficient.py

```diff
@@ -0,0 +1,56 @@
+import numpy as np
+import pytest
+from qibo import gates, models
+from scipy.optimize import minimize
+
+from qibochem.ansatz import hardware_efficient
+from qibochem.driver.molecule import Molecule
+from qibochem.measurement.expectation import expectation
+
+
+def test_hea_ansatz():
+    mol = Molecule([("H", (0.0, 0.0, 0.0)), ("H", (0.0, 0.0, 0.74804))])
+    mol.run_pyscf()
+    mol_classical_hf_energy = mol.e_hf
+    mol_sym_ham = mol.hamiltonian("s")
+
+    nlayers = 1
+    nqubits = mol.nso
+    ntheta = 2 * nqubits * nlayers
+    theta = np.zeros(ntheta)
+
+    hea_ansatz = hardware_efficient.hea(nlayers, nqubits)
+    qc = models.Circuit(nqubits)
+    qc.add(gates.X(_i) for _i in range(sum(mol.nelec)))
+    qc.add(hea_ansatz)
+    qc.set_parameters(theta)
+
+    hf_energy = expectation(qc, mol_sym_ham)
+    assert mol_classical_hf_energy == pytest.approx(hf_energy)
+
+
+def test_vqe_hea_ansatz():
+    def test_vqe_hea_ansatz_cost(parameters, circuit, hamiltonian):
+        circuit.set_parameters(parameters)
+        return expectation(circuit, hamiltonian)
+
+    mol = Molecule([("H", (0.0, 0.0, 0.0)), ("H", (0.0, 0.0, 0.75))])
+    mol.run_pyscf()
+    mol_classical_hf_energy = mol.e_hf
+    mol_sym_ham = mol.hamiltonian("s")
+
+    nlayers = 2
+    nqubits = mol.nso
+    ntheta = 2 * nqubits * nlayers
+    theta = np.full(ntheta, np.pi / 4)
+
+    hea_ansatz = hardware_efficient.hea(nlayers, nqubits)
+    qc = models.Circuit(nqubits)
+    qc.add(gates.X(_i) for _i in range(sum(mol.nelec)))
+    qc.add(hea_ansatz)
+    qc.set_parameters(theta)
+
+    vqe_object = minimize(test_vqe_hea_ansatz_cost, theta, args=(qc, mol_sym_ham), method="Powell")
+
+    vqe_hf_energy = vqe_object.fun
+    assert mol_classical_hf_energy == pytest.approx(vqe_hf_energy, 0.01)
```

Comment:
Just wondering, rather than testing against the HF energy again, would it be better for the VQE test to get the exact ground state energy instead? We can just test against a hardcoded value ( -1.13711707 ), no need to solve for it every time.

Reply:
IIRC this is at most what the HEA gives you, as we are only working within the space of unitaries that optimize each spin orbital. A systematic scan of all theta's within the parameter space should show the minimum being the HF energy. 

I don't think it will give you the FCI value, as it is optimized for hardware implementation and will not be as expressive as a chemically inspired ansatz such as UCCSD, which includes doubles excitations that get JW'd or BK'd and go into the ansatz thereafter. 

One thing that can be done, is to _not_ complete the SCF at the pyscf phase, and use this vqe process to obtain the correct parameters that give you the HF energy at SCF. 

What do you think?

Reply:
I do remember getting the FCI value when we did the DSO workshop last year, so I went back to look at it and did some testing to compare my code there to your code here.

So, I can't get the FCI value when `coupled_gate` is set to be a `CZ` gate, but can get it using `CNOT` gates instead. Try the following:

```
# Cost function
def test_vqe_hea_ansatz_cost(parameters, circuit, hamiltonian):
    circuit.set_parameters(parameters)
    return expectation(circuit, hamiltonian)

# Build Molecule
mol = Molecule([("H", (0.0, 0.0, 0.0)), ("H", (0.0, 0.0, 0.75))])
mol.run_pyscf()
mol_classical_hf_energy = mol.e_hf
mol_sym_ham = mol.hamiltonian("s")

nlayers = 2
nqubits = mol.nso

# Build HEA circuit using CZ gates
qc = models.Circuit(nqubits)
hea_cz_ansatz = hardware_efficient.hea(nlayers, nqubits)
qc.add(hea_cz_ansatz )

# Build HEA circuit using CNOT gates
qc2 = models.Circuit(nqubits)
hea_cnot_ansatz = hardware_efficient.hea(nlayers, nqubits, coupling_gates="CNOT")
qc2.add(hea_cnot_ansatz )

print("CZ coupling gates result:")
for _ in range(10):
    ntheta = len(qc.get_parameters())
    theta = np.random.rand(ntheta)
    vqe_object = minimize(test_vqe_hea_ansatz_cost, theta, args=(qc, mol_sym_ham), method="Powell")
    print(vqe_object.fun)
print()

print("CNOT coupling gates result:")
for _ in range(10):
    ntheta = len(qc2.get_parameters())
    theta = np.random.rand(ntheta)
    vqe_object = minimize(test_vqe_hea_ansatz_cost, theta, args=(qc2, mol_sym_ham), method="Powell")
    print(vqe_object.fun)
```

My CZ results are all about the HF value, while the CNOT results are either totally crap (-0.54...) or the FCI value.
Whether or not X gates are added before the HEA doesn't seem to affect the results much too.

Reply:
Indeed. 

It seems that for CZ coupling gates, we are limiting our solutions to within a narrow part of the Bloch sphere. 

I tried your CNOT method with 3 layers and I got the FCI solution for all 10 runs. 

Maybe we should omit the vqe test function for now?

Reply:
Personally, I would prefer to keep the VQE test and just use fixed initial values of `theta` for the optimization. Don't think we want the test to be too big, so just 3 layers would be good.

Reply:
Done in commit https://github.com/qiboteam/qibochem/commit/d21a1b40ff4a09ceecf1c89bb64acf7419cc6791